internal/provisionertask/provisioner_task.go

```diff
@@ -116,24 +116,35 @@ type RetryStrategy struct {
 	RetryCount int
 }
 
+// GetMachineInstanceInfoSetter provides the interface for setting the
+// instance info of a machine. It takes a machine provisioner API as input
+// so it can be used when we don't have a machine service available.
+type GetMachineInstanceInfoSetter func(machineProvisioner apiprovisioner.MachineProvisioner) func(
+	ctx context.Context,
+	id instance.Id, displayName string, nonce string, characteristics *instance.HardwareCharacteristics,
+	networkConfig []params.NetworkConfig, volumes []params.Volume,
+	volumeAttachments map[string]params.VolumeAttachmentInfo, charmProfiles []string,
+) error
```

Comment:
I so wish that params weren't at this level. This makes doing things a lot harder, as we're talking about API request/responses deep down inside a worker. 

The quick solution is to copy the param types into the provisionertask package and then convert them when we get them off the wire. Then the package __owns__ the types and we don't have to concern ourselves about an API here.

Reply:
I'd like to understand what's your plan. To be clear, we already have API dependencies (by IoC - `MachinesAPI` and `ControllerAPI`) in this same file, and we even already had the dependency to rpc params.

Are you striving for a provisioner task that doesn't interact with any APIs, solely worrying about business logic?
Or is it a pure technical concern about having a hard dependency on `rpc` from a worker so down the line?

I'd like us to discuss about it so we can better organize this worker, but in any of the two cases above, I think this can wait since it's not been introduced in this PR. What do you think?

Reply:
Both long term. Not to do now though

Reply:
OK, indeed this is exactly what I was bringing up when we discussed this off-line: I'd like the provisioner task to stop using any APIs and instead pass closures from the two provisioners when it needs to interact with the outside world. Something in the line of what we did with `RunAtomic()` for domain states.

Reply:
I'd rather they where interfaces that we just implemented